lib/vtls/darwinssl.c

```diff
@@ -1996,6 +2025,127 @@ static int verify_cert(const char *cafile, struct Curl_easy *data,
+static CURLcode pkp_pin_peer_pubkey(struct SessionHandle *data,
+                                    SSLContextRef ctx,
+                                    const char *pinnedpubkey)
+{  /* Scratch */
+  size_t pubkeylen, realpubkeylen, derHeaderLength = 24;
+  unsigned char *pubkey = NULL, *realpubkey = NULL, *derHeader = NULL;
+  CFDataRef publicKeyBits = NULL;
+
+  /* Result is returned to caller */
+  CURLcode result = CURLE_SSL_PINNEDPUBKEYNOTMATCH;
+
+  /* if a path wasn't specified, don't pin */
+  if(!pinnedpubkey)
+    return CURLE_OK;
+
+
+  if(!ctx)
+    return result;
+
+  do {
+    SecTrustRef trust;
+    OSStatus ret = SSLCopyPeerTrust(ctx, &trust);
+    if(ret != noErr || trust == NULL)
+        break;
+
+    SecKeyRef keyRef;
+
+#ifdef DARWIN_SSL_PINNEDPUBKEY_V1
+
+    /* 0 is leaf certificate, this is root:
+     * CFIndex certificateChainLen = SecTrustGetCertificateCount(trust);
+     */
+    SecCertificateRef cert = SecTrustGetCertificateAtIndex(trust, 0);
+    SecPolicyRef policy = SecPolicyCreateBasicX509();
+
+    SecTrustRef tempTrust;
+    SecTrustCreateWithCertificates(cert, policy, &tempTrust);
+    SecTrustEvaluate(tempTrust, NULL);
+    keyRef = SecTrustCopyPublicKey(tempTrust);
+    CFRelease(policy);
+    CFRelease(tempTrust);
+
+    publicKeyBits = SecKeyCopyExternalRepresentation(keyRef, NULL);
+    CFRelease(keyRef);
+    CFRelease(trust);
+    if(publicKeyBits == NULL)
+        break;
+
+#elif DARWIN_SSL_PINNEDPUBKEY_V2
+
+    keyRef = SecTrustCopyPublicKey(trust);
+    CFRelease(trust);
+
+    OSStatus success = SecItemExport(keyRef, kSecFormatOpenSSL, 0, NULL,
+                                     &publicKeyBits);
+    CFRelease(keyRef);
+    if(success != errSecSuccess || publicKeyBits == NULL)
+        break;
+
+#endif /* DARWIN_SSL_PINNEDPUBKEY_V2 */
+
+    pubkeylen = CFDataGetLength(publicKeyBits);
+    pubkey = CFDataGetBytePtr(publicKeyBits);
+
+    switch(pubkeylen) {
+      case 526:
+        /* 4096 bit RSA pubkeylen == 526 */
+        derHeader = rsa4096DerHeader;
+        break;
+      case 270:
+        /* 2048 bit RSA pubkeylen == 270 */
+        derHeader = rsa2048DerHeader;
+        break;
+#ifdef DARWIN_SSL_PINNEDPUBKEY_V1
+      case 65:
+        /* ecDSA secp256r1 pubkeylen == 65 */
+        derHeader = ecDsaSecp256r1DerHeader;
+        derHeaderLength = 26;
+        break;
+      case 97:
+        /* ecDSA secp384r1 pubkeylen == 97 */
+        derHeader = ecDsaSecp384r1DerHeader;
+        derHeaderLength = 23;
+        break;
+      default:
+        infof(data, "SSL: unhandled public key length: %d\n", pubkeylen);
+#elif DARWIN_SSL_PINNEDPUBKEY_V2
+      default:
+        /* ecDSA secp256r1 pubkeylen == 91 header already included?
+         * ecDSA secp384r1 header already included too
+         * we assume rest of algorithms do same, so do nothing
+         */
+        result = Curl_pin_peer_pubkey(data, pinnedpubkey, pubkey,
+                                    pubkeylen);
```

Comment:
Yes, but I did it below with the if(derHeader == NULL) break; statement, I guess breaking on labels isn't supported in C89 ?